src/ngx_http_lua_ssl.c

```diff
@@ -34,4 +34,129 @@ ngx_http_lua_ssl_init(ngx_log_t *log)
 }
 
 
+#ifndef NGX_LUA_NO_FFI_API
+
+
+void *
+ngx_http_lua_ffi_ssl_ctx_init(ngx_uint_t protocols, char **err)
+{
+    ngx_ssl_t                ssl;
+
+    ssl.log = ngx_cycle->log;
+    if (ngx_ssl_create(&ssl, protocols, NULL) != NGX_OK) {
+        *err = "failed to create ssl ctx";
+        ngx_log_error(NGX_LOG_ERR, ssl.log, 0, *err);
+        return NULL;
+    }
+
+    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, ssl.log, 0,
+                   "lua ssl ctx init: %p:%d", ssl.ctx, ssl.ctx->references);
+
+    return ssl.ctx;
+}
+
+
+int
+ngx_http_lua_ffi_ssl_ctx_set_cert(void *cdata_ctx, void *cdata_cert, char **err)
+{
+#ifdef LIBRESSL_VERSION_NUMBER
+
+    *err = "LibreSSL not supported";
+    return NGX_ERROR;
+
+#else
+
+#   if OPENSSL_VERSION_NUMBER < 0x1000205fL
+
+    *err = "at least OpenSSL 1.0.2e required but found " OPENSSL_VERSION_TEXT;
+    return NGX_ERROR;
+
+#   else
+
+    X509              *x509 = NULL;
+    SSL_CTX           *ssl_ctx = cdata_ctx;
+    STACK_OF(X509)    *cert = cdata_cert;
+
+#ifdef OPENSSL_IS_BORINGSSL
+    size_t             i;
+#else
+    int                i;
+#endif
+
+    if (sk_X509_num(cert) < 1) {
+        *err = "sk_X509_num() failed";
+        ngx_ssl_error(NGX_LOG_ERR, ngx_cycle->log, 0, *err);
+        return NGX_ERROR;
+    }
+
+    x509 = sk_X509_value(cert, 0);
+    if (x509 == NULL) {
+        *err = "sk_X509_value() failed";
+        ngx_ssl_error(NGX_LOG_ERR, ngx_cycle->log, 0, *err);
```

Comment:
@doujiang24 Maybe the `err_len` is no need. I check the API `char *ERR_error_string(unsigned long e, char *buf);` It says:
>ERR_error_string() generates a human-readable string representing the error code e, and places it at buf. buf must be at least 120 bytes long. If buf is NULL , the error string is placed in a static buffer. 

So pointing the static buffer is not bad IMO as the following:

```c
   if (SSL_CTX_use_PrivateKey(ssl_ctx, key) == 0) {
        e = ERR_get_error();
        if (e == 0) {
            *err = "SSL_CTX_use_PrivateKey() failed";

        } else {
            *err = ERR_error_string(e, NULL);
        }

        return NGX_ERROR;
    }
```

@doujiang24  @agentzh How about this?